File: archive/cdktf-ts/Pr824/lib/security.ts
// lib/security.ts
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { TerraformOutput } from 'cdktf';
import { Construct } from 'constructs';
import { name } from './utils/naming';

export interface SecurityProps {
  provider: AwsProvider;
  environment: string;
  region: string;
  vpcId: string;

  openAlbHttp?: boolean; // default true
  openAlbHttps?: boolean; // default true
  adminCidr?: string; // e.g., "203.0.113.0/24"
  appPort?: number; // default 80
  enableSshToApp?: boolean; // default false
}

export class Security extends Construct {
  public readonly albSgId: string;
  public readonly appSgId: string;
  public readonly rdsSgId: string;

  constructor(scope: Construct, id: string, props: SecurityProps) {
    super(scope, id);

    const env = props.environment;
    const region = props.region;
    const appPort = props.appPort ?? 80;
    const openAlbHttp = props.openAlbHttp ?? true;
    const openAlbHttps = props.openAlbHttps ?? true;
    const enableSshToApp = props.enableSshToApp ?? false;

    // ALB SG
    const albSg = new SecurityGroup(this, 'alb-sg', {
      name: name(env, 'alb-sg', region),
      description: 'Security group for Application Load Balancer',
      vpcId: props.vpcId,
      egress: [
        { fromPort: 0, toPort: 0, protocol: '-1', cidrBlocks: ['0.0.0.0/0'] },
      ],
      ingress: [
        ...(openAlbHttp
          ? [
              {
                fromPort: 80,
                toPort: 80,
                protocol: 'tcp',
                cidrBlocks: ['0.0.0.0/0'],
              },
            ]
          : []),
        ...(openAlbHttps
          ? [
              {
                fromPort: 443,
                toPort: 443,
                protocol: 'tcp',
                cidrBlocks: ['0.0.0.0/0'],
              },
            ]
          : []),
      ] as any[],
      provider: props.provider,
      tags: { Name: name(env, 'alb-sg', region) },
    });

    // App SG ingress (allow from ALB SG on app port; optional SSH from admin CIDR)
    const appIngress: any[] = [
      {
        fromPort: appPort,
        toPort: appPort,
        protocol: 'tcp',
        securityGroups: [albSg.id],
      },
    ];

    if (enableSshToApp) {
      const adminCidr = props.adminCidr || process.env.ADMIN_CIDR || '';
      if (adminCidr.trim().length > 0) {
        appIngress.push({
          fromPort: 22,
          toPort: 22,
          protocol: 'tcp',
          cidrBlocks: [adminCidr],
        });
      }
    }

    const appSg = new SecurityGroup(this, 'app-sg', {
      name: name(env, 'app-sg', region),
      description: 'Security group for application instances',
      vpcId: props.vpcId,
      ingress: appIngress,
      egress: [
        { fromPort: 0, toPort: 0, protocol: '-1', cidrBlocks: ['0.0.0.0/0'] },
      ],
      provider: props.provider,
      tags: { Name: name(env, 'app-sg', region) },
    });

    // RDS SG (only from app SG on 5432)
    const rdsSg = new SecurityGroup(this, 'rds-sg', {
      name: name(env, 'rds-sg', region),
      description: 'Security group for RDS Postgres',
      vpcId: props.vpcId,
      ingress: [
        {
          fromPort: 5432,
          toPort: 5432,
          protocol: 'tcp',
          securityGroups: [appSg.id],
        },
      ] as any[],
      egress: [
        { fromPort: 0, toPort: 0, protocol: '-1', cidrBlocks: ['0.0.0.0/0'] },
      ],
      provider: props.provider,
      tags: { Name: name(env, 'rds-sg', region) },
    });

    this.albSgId = albSg.id;
    this.appSgId = appSg.id;
    this.rdsSgId = rdsSg.id;

    new TerraformOutput(this, 'alb_sg_id', { value: this.albSgId });
    new TerraformOutput(this, 'app_sg_id', { value: this.appSgId });
    new TerraformOutput(this, 'rds_sg_id', { value: this.rdsSgId });
  }
}

File: archive/pulumi-py/Pr7064/lib/security.py
"""Security Groups"""
import pulumi
import pulumi_aws as aws
from typing import Dict

def create_security_groups(
    environment_suffix: str,
    vpc_id: pulumi.Output[str],
    vpc_cidr: str,
    tags: Dict[str, str]
) -> Dict[str, aws.ec2.SecurityGroup]:
    """Create security groups for all components"""

    # ALB Security Group
    alb_sg = aws.ec2.SecurityGroup(
        f"payment-alb-sg-{environment_suffix}",
        vpc_id=vpc_id,
        description="Security group for Application Load Balancer",
        ingress=[
            aws.ec2.SecurityGroupIngressArgs(
                protocol="tcp",
                from_port=80,
                to_port=80,
                cidr_blocks=["0.0.0.0/0"],
                description="HTTP from internet"
            ),
            aws.ec2.SecurityGroupIngressArgs(
                protocol="tcp",
                from_port=443,
                to_port=443,
                cidr_blocks=["0.0.0.0/0"],
                description="HTTPS from internet"
            )
        ],
        egress=[
            aws.ec2.SecurityGroupEgressArgs(
                protocol="-1",
                from_port=0,
                to_port=0,
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound"
            )
        ],
        tags={**tags, "Name": f"payment-alb-sg-{environment_suffix}"}
    )

    # Application Security Group
    app_sg = aws.ec2.SecurityGroup(
        f"payment-app-sg-{environment_suffix}",
        vpc_id=vpc_id,
        description="Security group for ECS application containers",
        ingress=[
            aws.ec2.SecurityGroupIngressArgs(
                protocol="tcp",
                from_port=8080,
                to_port=8080,
                security_groups=[alb_sg.id],
                description="Allow traffic from ALB"
            )
        ],
        egress=[
            aws.ec2.SecurityGroupEgressArgs(
                protocol="-1",
                from_port=0,
                to_port=0,
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound"
            )
        ],
        tags={**tags, "Name": f"payment-app-sg-{environment_suffix}"}
    )

    # Database Security Group
    database_sg = aws.ec2.SecurityGroup(
        f"payment-db-sg-{environment_suffix}",
        vpc_id=vpc_id,
        description="Security group for RDS Aurora database",
        ingress=[
            aws.ec2.SecurityGroupIngressArgs(
                protocol="tcp",
                from_port=5432,
                to_port=5432,
                security_groups=[app_sg.id],
                description="PostgreSQL from application"
            )
        ],
        egress=[
            aws.ec2.SecurityGroupEgressArgs(
                protocol="-1",
                from_port=0,
                to_port=0,
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound"
            )
        ],
        tags={**tags, "Name": f"payment-db-sg-{environment_suffix}"}
    )

    # Cache Security Group
    cache_sg = aws.ec2.SecurityGroup(
        f"payment-cache-sg-{environment_suffix}",
        vpc_id=vpc_id,
        description="Security group for ElastiCache Redis",
        ingress=[
            aws.ec2.SecurityGroupIngressArgs(
                protocol="tcp",
                from_port=6379,
                to_port=6379,
                security_groups=[app_sg.id],
                description="Redis from application"
            )
        ],
        egress=[
            aws.ec2.SecurityGroupEgressArgs(
                protocol="-1",
                from_port=0,
                to_port=0,
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound"
            )
        ],
        tags={**tags, "Name": f"payment-cache-sg-{environment_suffix}"}
    )

    return {
        "alb_sg": alb_sg,
        "app_sg": app_sg,
        "database_sg": database_sg,
        "cache_sg": cache_sg
    }
